src/ocean/util/log/Stats.d

```diff
@@ -650,3 +656,402 @@ unittest
+        int result;
+
+        foreach (line; this.lines)
+        {
+            Const!(StatsLine) stats_line = StatsLine(line);
+            result = dg(stats_line);
+
+            if ( result != 0 )
+            {
+                break;
+            }
+        }
+
+        return result;
+    }
+
+    /******************************************************************************
+
+        ditto
+
+    ******************************************************************************/
+
+    public int opApply ( int delegate(ref size_t index, ref Const!(StatsLine)) dg )
+    {
+        int result;
+        size_t index;
+
+        foreach (line; this.lines)
+        {
+            Const!(StatsLine) stats_line = StatsLine(line);
+            result = dg(index, stats_line);
+            index++;
+
+            if ( result != 0 )
+            {
+                break;
+            }
+        }
+
+        return result;
+    }
+}
+
+/// Read a list of stats
+unittest
+{
+    auto data = new Array(
+        "2018-09-12 10:03:07,598 2018-09-12 10:03:07,598 cpu_usage:1 memory:3\n" ~
+        "2018-09-12 10:03:07,598 2018-09-12 10:03:07,598 cpu_usage:2 memory:4".dup);
+
+    auto reader = new StatsLogReader(data);
+
+    /// Iteration without index
+    size_t index;
+    foreach (line; reader)
+    {
+        if (index == 0)
+        {
+            test!("==")(line["cpu_usage"], "1");
+            test!("==")(line["memory"], "3");
+        }
+        else
+        {
+            test!("==")(line["cpu_usage"], "2");
+            test!("==")(line["memory"], "4");
+        }
+        index++;
+    }
+
+    test!("==")(index, 2);
+
+    /// Iteration with index
+    data = new Array(
+        "2018-09-12 10:03:07,598 2018-09-12 10:03:07,598 cpu_usage:1 memory:3\n" ~
+        "2018-09-12 10:03:07,598 2018-09-12 10:03:07,598 cpu_usage:2 memory:4".dup);
+    reader = new StatsLogReader(data);
+    index = 0;
+    foreach (i, line; reader)
+    {
+        if (i == 0)
+        {
+            test!("==")(line["cpu_usage"], "1");
+            test!("==")(line["memory"], "3");
+        }
+        else
+        {
+            test!("==")(line["cpu_usage"], "2");
+            test!("==")(line["memory"], "4");
+        }
+
+        index = i;
+    }
+
+    test!("==")(index, 1);
+}
+
+/// Get the last line
+unittest
+{
+    auto data = new Array(
+        "2018-09-12 10:03:07,598 2018-09-12 10:03:07,598 cpu_usage:1 memory:3\n" ~
+        "2018-09-12 10:03:07,598 2018-09-12 10:03:07,598 cpu_usage:2 memory:4".dup);
+
+    auto reader = new StatsLogReader(data);
+    auto line = reader.last();
+
+    test!("==")(line["cpu_usage"], "2");
+    test!("==")(line["memory"], "4");
+
+    /// It should raise an error for an empty string
+    bool has_exception;
+
+    try
+    {
+        data = new Array("".dup);
+        reader = new StatsLogReader(data);
+        reader.last();
+    }
+    catch(Exception e)
+    {
+        test!("==")(e.message, "The stats are empty");
```

Comment:
Why not simply `testThrown`?

Reply:
because `testThrown` can not check the error message. I updated the code to use `testThrown` anyway... It looks like there are no tests that check for the error messages

Reply:
Yeah we don't normally care about exact error message and it is not worth extra lines of boilerplate. If specific exception kind is crucial, it should be denoted by creating dedicated exception class and doing `testThrown!(MyExceptionClass)` instead.